File: .tmp/parse_identification.py
#!/usr/bin/env python3
"""Parse the logs CSV and extract identification-phase LLM calls."""
import csv
import json
import os
import sys
from pathlib import Path

# Ensure we can handle the very long fields
csv.field_size_limit(sys.maxsize if hasattr(sys, 'maxsize') else 2**31 - 1)
try:
    csv.field_size_limit(2**31 - 1)
except OverflowError:
    csv.field_size_limit(2**30)

CSV_PATH = r"e:/Work/ebike/repos/ebike-backend/docs/thread-extraction/736b1b30-d7da-4ddf-8504-a6957b64c789-logs.csv"
OUT_PATH = r"e:/Work/ebike/repos/ebike-backend/.tmp/identification-subtrees.json"

IDENTIFICATION_PROMPT_PREFIX = "You identify which products each Reddit comment references."
RECEIVED_MARKER = "AI chat response received [openai:gpt-5.4-mini]"


def find_json_payload(line: str):
    """Find the JSON metadata object in a Loki log line.

    The log line typically looks like a free-form string with an embedded JSON object.
    We try to find the largest balanced { ... } substring that parses as JSON.
    """
    # Find candidate starts (likely the metadata blob comes after the message text)
    candidates = []
    for start in range(len(line)):
        if line[start] != '{':
            continue
        # try to balance braces
        depth = 0
        in_str = False
        escape = False
        for end in range(start, len(line)):
            ch = line[end]
            if escape:
                escape = False
                continue
            if ch == '\\':
                escape = True
                continue
            if ch == '"':
                in_str = not in_str
                continue
            if in_str:
                continue
            if ch == '{':
                depth += 1
            elif ch == '}':
                depth -= 1
                if depth == 0:
                    candidate = line[start:end + 1]
                    try:
                        obj = json.loads(candidate)
                        if isinstance(obj, dict):
                            candidates.append(obj)
                    except json.JSONDecodeError:
                        pass
                    break
    if not candidates:
        return None
    # Pick the candidate with most keys (likely the full metadata)
    return max(candidates, key=lambda d: len(d.keys()) if isinstance(d, dict) else 0)


def extract_identification_calls():
    rows = []
    with open(CSV_PATH, 'r', encoding='utf-8', newline='') as f:
        reader = csv.DictReader(f)
        for row in reader:
            line = row.get('Line', '')
            if not line:
                continue
            if RECEIVED_MARKER not in line:
                continue
            # Quick filter: ensure prompt prefix appears in the line at all
            if IDENTIFICATION_PROMPT_PREFIX not in line:
                continue
            payload = find_json_payload(line)
            if not payload:
                continue
            messages = payload.get('messages')
            if not isinstance(messages, list) or len(messages) < 2:
                continue
            sys_msg = messages[0]
            sys_content = sys_msg.get('content') if isinstance(sys_msg, dict) else None
            if not isinstance(sys_content, str):
                continue
            if not sys_content.startswith(IDENTIFICATION_PROMPT_PREFIX):
                continue
            user_msg = messages[1]
            user_content = user_msg.get('content') if isinstance(user_msg, dict) else None
            response = payload.get('response', '')
            chat_id = payload.get('chatId') or payload.get('chat_id') or ''
            subtree_id = payload.get('subtreeId') or payload.get('subtree_id') or ''
            ts_ns = row.get('tsNs') or ''
            rows.append({
                'tsNs': ts_ns,
                'chatId': chat_id,
                'subtreeId': subtree_id,
                'userPrompt': user_content if isinstance(user_content, str) else json.dumps(user_content),
                'response': response if isinstance(response, str) else json.dumps(response),
            })
    # Sort chronologically (oldest first). tsNs is a numeric string.
    def sort_key(r):
        try:
            return int(r['tsNs'])
        except Exception:
            return 0
    rows.sort(key=sort_key)
    return rows


def main():
    calls = extract_identification_calls()
    out = {
        'subtreeCount': len(calls),
        'subtrees': [
            {
                'chatId': c['chatId'],
                'subtreeId': c['subtreeId'],
                'userPrompt': c['userPrompt'],
                'response': c['response'],
            }
            for c in calls
        ],
    }
    Path(os.path.dirname(OUT_PATH)).mkdir(parents=True, exist_ok=True)
    with open(OUT_PATH, 'w', encoding='utf-8') as f:
        json.dump(out, f, ensure_ascii=False, indent=2)
    print(f"Found {len(calls)} identification calls")
    print(f"Wrote: {OUT_PATH}")
    if calls:
        first = calls[0]
        print(f"\nFirst subtreeId: {first['subtreeId']}")
        print(f"First chatId: {first['chatId']}")
        prompt = first['userPrompt']
        # Find Comments section and show first ~2000 chars
        idx = prompt.find('## Comments')
        if idx >= 0:
            snippet = prompt[idx:idx + 1500]
        else:
            snippet = prompt[:1500]
        print("\n--- First subtree comments section snippet ---")
        print(snippet)


if __name__ == '__main__':
    main()

File: .tmp/parse_identification.mjs
// Stream-parse the logs CSV and extract identification-phase LLM calls.
import fs from "node:fs";
import readline from "node:readline";
import path from "node:path";

const CSV_PATH =
  "e:/Work/fittkereso/repos/fittkereso-backend/docs/thread-extraction/736b1b30-d7da-4ddf-8504-a6957b64c789-logs.csv";
const OUT_PATH =
  "e:/Work/fittkereso/repos/fittkereso-backend/.tmp/identification-subtrees.json";

const IDENTIFICATION_PROMPT_PREFIX =
  "You identify which products each Reddit comment references.";
const RECEIVED_MARKER = "AI chat response received [openai:gpt-5.4-mini]";

/**
 * Parse a single CSV row, accounting for RFC-4180 doubled-quote escaping ("").
 * Returns an array of fields. Assumes the row is one logical line (no embedded newlines —
 * the CSV here is one logical entry per file line).
 */
function parseCsvRow(line) {
  const fields = [];
  let i = 0;
  let cur = "";
  let inQuotes = false;
  while (i < line.length) {
    const ch = line[i];
    if (inQuotes) {
      if (ch === '"') {
        if (line[i + 1] === '"') {
          cur += '"';
          i += 2;
          continue;
        }
        inQuotes = false;
        i += 1;
        continue;
      }
      cur += ch;
      i += 1;
      continue;
    }
    // not in quotes
    if (ch === '"') {
      inQuotes = true;
      i += 1;
      continue;
    }
    if (ch === ",") {
      fields.push(cur);
      cur = "";
      i += 1;
      continue;
    }
    cur += ch;
    i += 1;
  }
  fields.push(cur);
  return fields;
}

/**
 * Find the largest balanced JSON object inside a string and return the parsed object
 * with the most keys. Returns null if none found.
 */
function findBestJsonObject(text) {
  let best = null;
  let bestKeys = -1;
  for (let start = 0; start < text.length; start++) {
    if (text[start] !== "{") continue;
    let depth = 0;
    let inStr = false;
    let escape = false;
    for (let end = start; end < text.length; end++) {
      const ch = text[end];
      if (escape) {
        escape = false;
        continue;
      }
      if (ch === "\\") {
        escape = true;
        continue;
      }
      if (ch === '"') {
        inStr = !inStr;
        continue;
      }
      if (inStr) continue;
      if (ch === "{") depth += 1;
      else if (ch === "}") {
        depth -= 1;
        if (depth === 0) {
          const candidate = text.slice(start, end + 1);
          try {
            const obj = JSON.parse(candidate);
            if (obj && typeof obj === "object" && !Array.isArray(obj)) {
              const keyCount = Object.keys(obj).length;
              if (keyCount > bestKeys) {
                best = obj;
                bestKeys = keyCount;
              }
            }
          } catch {
            // not valid JSON; ignore
          }
          break;
        }
      }
    }
  }
  return best;
}

async function run() {
  const stream = fs.createReadStream(CSV_PATH, { encoding: "utf-8" });
  const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });

  let header = null;
  let lineColIdx = -1;
  let tsColIdx = -1;
  const calls = [];

  let lineNum = 0;
  for await (const rawLine of rl) {
    lineNum += 1;
    if (!rawLine) continue;
    if (header === null) {
      header = parseCsvRow(rawLine);
      lineColIdx = header.indexOf("Line");
      tsColIdx = header.indexOf("tsNs");
      continue;
    }
    // Cheap pre-filter: skip rows we obviously don't want.
    if (!rawLine.includes(RECEIVED_MARKER)) continue;
    if (!rawLine.includes(IDENTIFICATION_PROMPT_PREFIX)) continue;

    const fields = parseCsvRow(rawLine);
    if (lineColIdx < 0 || lineColIdx >= fields.length) continue;
    const line = fields[lineColIdx];
    const tsNs =
      tsColIdx >= 0 && tsColIdx < fields.length ? fields[tsColIdx] : "";

    if (!line.includes(RECEIVED_MARKER)) continue;
    if (!line.includes(IDENTIFICATION_PROMPT_PREFIX)) continue;

    const payload = findBestJsonObject(line);
    if (!payload) continue;
    const messages = payload.messages;
    if (!Array.isArray(messages) || messages.length < 2) continue;
    const sysContent = messages[0]?.content;
    if (typeof sysContent !== "string") continue;
    if (!sysContent.startsWith(IDENTIFICATION_PROMPT_PREFIX)) continue;
    const userContent = messages[1]?.content;
    const userPrompt =
      typeof userContent === "string"
        ? userContent
        : JSON.stringify(userContent);
    const response =
      typeof payload.response === "string"
        ? payload.response
        : payload.response != null
          ? JSON.stringify(payload.response)
          : "";
    const chatId = payload.chatId || payload.chat_id || "";
    const subtreeId = payload.subtreeId || payload.subtree_id || "";

    calls.push({ tsNs, chatId, subtreeId, userPrompt, response });
  }

  // Sort chronologically (oldest first); tsNs is numeric (BigInt-safe compare via string length first).
  calls.sort((a, b) => {
    const aa = a.tsNs || "0";
    const bb = b.tsNs || "0";
    if (aa.length !== bb.length) return aa.length - bb.length;
    if (aa < bb) return -1;
    if (aa > bb) return 1;
    return 0;
  });

  const out = {
    subtreeCount: calls.length,
    subtrees: calls.map((c) => ({
      chatId: c.chatId,
      subtreeId: c.subtreeId,
      userPrompt: c.userPrompt,
      response: c.response,
    })),
  };

  fs.mkdirSync(path.dirname(OUT_PATH), { recursive: true });
  fs.writeFileSync(OUT_PATH, JSON.stringify(out, null, 2), "utf-8");

  console.log(`Found ${calls.length} identification calls`);
  console.log(`Wrote: ${OUT_PATH}`);
  if (calls.length > 0) {
    const first = calls[0];
    console.log(`\nFirst subtreeId: ${first.subtreeId}`);
    console.log(`First chatId: ${first.chatId}`);
    const idx = first.userPrompt.indexOf("## Comments");
    const snippet =
      idx >= 0
        ? first.userPrompt.slice(idx, idx + 1800)
        : first.userPrompt.slice(0, 1800);
    console.log("\n--- First subtree Comments section snippet ---");
    console.log(snippet);
  }
}

run().catch((err) => {
  console.error(err);
  process.exit(1);
});
